{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Goal: Classify Genetic Variations using Text from Clinical Evidence\n",
    "A team at Memorial Sloan Kettering Cancer Center (MSKCC) spends counless hours every year manually reviewing evidence to classify genetic mutations. Once the most important variants have been identified new therapies can be developed. Having a machine learning algorithm to automatically classify genetic variations would allow the development of therapies much sooner. This competition is a challenge to develop classification models which analyze abstracts of medical articles and, based on their content accurately determine mutation effect (9 classes) of the genes discussed in them. For more information, please refer to the Kaggle competition: https://www.kaggle.com/c/msk-redefining-cancer-treatment#description. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is an article that also gives context to a project such as this and the impact could have: https://www.forbes.com/sites/matthewherper/2017/06/03/a-new-cancer-drug-helped-almost-everyone-who-took-it-almost-heres-what-it-teaches-us/#12aad9136b25. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This task is complex and hard to elvaluate the cross val score due to the data only representing some of the genes in the dataset. In the end, if my score is not great, I hope to have been able to better understand the use of machine learning in genomic analysis and explore its limitations. \n",
    "\n",
    "\n",
    "Also, there was a 2nd phase to this competition due to the labels for the test data being released by a third party before the competition closed. I may want to combine the training data from both phases to train the model if I am not able to get a good score from just the initial data set. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Target: 9 classes that represent different levels of whether or not the variant is a driver or passenger mutation.\n",
    "These are the class labels from OncoKB, but using this text in the competition is prohibited. \n",
    "1. Likely Loss-of-function\n",
    "2. Likely Gain-of-function\n",
    "3. Neutral\n",
    "4. Loss-of-function\n",
    "5. Likely Neutral\n",
    "6. Inconclusive\n",
    "7. Gain-of-function\n",
    "8. Likely Switch-of-function\n",
    "9. Switch-of-function\n",
    "\n",
    "Using COSMIC and Genotype-Phenotype associations are not allowed due to the point of this project being modeling text.\n",
    "\n",
    "You are not allowed to use conservation scores, or predictors of deleterious or tolerated mutations (such as Polyphen) because these tools also try to predict the effect of a mutation (which is very close to the goal of this competition) but they don't use text to do it"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Outside resources that have been approved by kaggle for use in this competition but not necessary:\n",
    " -  Gene–disease associations: http://ctdbase.org/\n",
    " -  Text data from:\n",
    "    - Encyclopedia_of_Molecular_Cell_Biology_and_Molecular_Medicine_16_volumes_Wiley_2006_\n",
    "    - Cell-Molecular-Biology-Concepts-Experiments 7th edition\n",
    " - MESH topics, as well as extended information about those articles\n",
    " - Pre-trained Bio-NER taggers (May need to map the variations to RSIDs and search related articles in PubMed or PMC):\n",
    "    - tmVar https://www.ncbi.nlm.nih.gov/research/bionlp/Tools/tmvar/\n",
    "    - TaggerOne https://www.ncbi.nlm.nih.gov/research/bionlp/Tools/taggerone/\n",
    "    - GNormPlus https://www.ncbi.nlm.nih.gov/research/bionlp/Tools/gnormplus/\n",
    " - Pathway and protein information of each gene (e.g. GO).\n",
    " - Pubmed dictionaries of genes names.\n",
    " - Mygene api (http://docs.mygene.info/en/v3/doc/data.html) It contains a lot of information about genes (such as family, overall description, etc). \n",
    " - Gene expression from The Human Protein Atlas project ( https://www.ebi.ac.uk/gxa/experiments/E-PROT-3/Downloads )\n",
    " - The aminoacid sequences for the genes in the dataset, in FASTA.\n",
    " - Other papers from PubMed in addition to the training_text in order to pre-train NLP model\n",
    " - Genia, an annotated treebank for parsing biomed papers(http://www.geniaproject.org/)\n",
    " - You can use variations themselves as features. For example a 'Deletion' in a gene may be more likely to be 1. Similar case could be made for amino acid changes. Some amino acid changes are more deleterious than others.\n",
    " - Google's trained Word2Vec models and pretrained models from bio.nlplab.org\n",
    " - Sequence data, mutation positition data and protein structure data from Uniport\n",
    " - allele frequency information from gnomAD or ExAC\n",
    " - controlled vocabularies to guide Named Entity Recognition modles. Examples: such as ICD, SNOMED CT,NCI Thesaurus, CPT, MedDRA, SNOMED CT\n",
    " \n",
    " Note from Kaggle: \"We knew the task was very difficult, so Iker from MSK suggested that we allow external data, in case anyone wants to build a knowledge base for gene mutations.\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from collections import Counter\n",
    "\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "train_var = pd.read_csv('Data/training_variants.zip')\n",
    "train_txt = pd.read_csv('Data/training_text.zip', sep=\"\\|\\|\", engine=\"python\", skiprows=1, names=[\"ID\", \"Text\"])\n",
    "test_var = pd.read_csv('Data/test_variants.zip')\n",
    "test_txt = pd.read_csv('Data/test_text.zip', sep=\"\\|\\|\", engine=\"python\", skiprows=1, names=[\"ID\", \"Text\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "# EDA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Training Variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3321, 4)\n",
      "ID            int64\n",
      "Gene         object\n",
      "Variation    object\n",
      "Class         int64\n",
      "dtype: object\n",
      "ID           0\n",
      "Gene         0\n",
      "Variation    0\n",
      "Class        0\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "      <th>Class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>FAM58A</td>\n",
       "      <td>Truncating Mutations</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>CBL</td>\n",
       "      <td>W802*</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>CBL</td>\n",
       "      <td>Q249E</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>CBL</td>\n",
       "      <td>N454D</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>CBL</td>\n",
       "      <td>L399V</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID    Gene             Variation  Class\n",
       "0   0  FAM58A  Truncating Mutations      1\n",
       "1   1     CBL                 W802*      2\n",
       "2   2     CBL                 Q249E      2\n",
       "3   3     CBL                 N454D      3\n",
       "4   4     CBL                 L399V      4"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(train_var.shape)\n",
    "print(train_var.dtypes)\n",
    "print(train_var.isnull().sum())\n",
    "train_var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>BRCA1</th>\n",
       "      <td>264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53</th>\n",
       "      <td>163</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EGFR</th>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTEN</th>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA2</th>\n",
       "      <td>125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIT</th>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRAF</th>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERBB2</th>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALK</th>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PDGFRA</th>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PIK3CA</th>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDKN2A</th>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGFR2</th>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLT3</th>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSC2</th>\n",
       "      <td>47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MTOR</th>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRAS</th>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MAP2K1</th>\n",
       "      <td>43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VHL</th>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RET</th>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGFR3</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MLH1</th>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SMAD4</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MET</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAK2</th>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOTCH1</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AKT1</th>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPN11</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABL1</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ROS1</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IKBKE</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AXL</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FANCC</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF1AX</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGF4</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRSF2</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KDM5A</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VEGFA</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGF3</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LATS1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCF3</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FGF19</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDKN2C</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PAX8</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GNAQ</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HLA-B</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RNF43</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BCL2</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MDM4</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SDHB</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HIST1H1C</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MEN1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLT1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SHOC2</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RICTOR</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM58A</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2B</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PMS1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOXO1</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLF4</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>264 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene\n",
       "BRCA1      264\n",
       "TP53       163\n",
       "EGFR       141\n",
       "PTEN       126\n",
       "BRCA2      125\n",
       "KIT         99\n",
       "BRAF        93\n",
       "ERBB2       69\n",
       "ALK         69\n",
       "PDGFRA      60\n",
       "PIK3CA      56\n",
       "CDKN2A      52\n",
       "FGFR2       50\n",
       "FLT3        49\n",
       "TSC2        47\n",
       "MTOR        45\n",
       "KRAS        44\n",
       "MAP2K1      43\n",
       "VHL         41\n",
       "RET         40\n",
       "FGFR3       39\n",
       "MLH1        35\n",
       "SMAD4       33\n",
       "MET         33\n",
       "JAK2        33\n",
       "NOTCH1      31\n",
       "AKT1        28\n",
       "PTPN11      26\n",
       "ABL1        26\n",
       "ROS1        26\n",
       "...        ...\n",
       "IKBKE        1\n",
       "AXL          1\n",
       "FANCC        1\n",
       "EIF1AX       1\n",
       "FGF4         1\n",
       "SRSF2        1\n",
       "KDM5A        1\n",
       "VEGFA        1\n",
       "FGF3         1\n",
       "LATS1        1\n",
       "TCF3         1\n",
       "FGF19        1\n",
       "CDKN2C       1\n",
       "PAX8         1\n",
       "GNAQ         1\n",
       "HLA-B        1\n",
       "RNF43        1\n",
       "BCL2         1\n",
       "MDM4         1\n",
       "SDHB         1\n",
       "HIST1H1C     1\n",
       "MEN1         1\n",
       "FLT1         1\n",
       "SHOC2        1\n",
       "RICTOR       1\n",
       "FAM58A       1\n",
       "KMT2B        1\n",
       "PMS1         1\n",
       "FOXO1        1\n",
       "KLF4         1\n",
       "\n",
       "[264 rows x 1 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(train_var['Gene'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Truncating Mutations</th>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Deletion</th>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Amplification</th>\n",
       "      <td>71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fusions</th>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Overexpression</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12V</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61R</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T58I</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E17K</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61L</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61H</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K117N</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y42C</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Promoter Hypermethylation</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q209L</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ETV6-NTRK3 Fusion</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q61K</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A146V</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R170W</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M1R</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EWSR1-ETV1 Fusion</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I31M</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y64A</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMPRSS2-ETV1 Fusion</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T286A</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G12S</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G35R</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P34R</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q22K</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S222D</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y155C</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D404G</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y35C</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R156C</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S1206Y</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1022E</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E462G</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I1171T</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I157T</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N116H</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R897Q</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D402Y</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S723F</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V379I</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E709K</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D1091N</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N581S</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A636P</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C24Y</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R625L</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1066V</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L1301R</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K459_S460delinsN</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L858Q</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P179L</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L485F</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y65C</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W557_V560del</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S240R</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STRN-PDGFRA Fusion</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2996 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Variation\n",
       "Truncating Mutations              93\n",
       "Deletion                          74\n",
       "Amplification                     71\n",
       "Fusions                           34\n",
       "Overexpression                     6\n",
       "G12V                               4\n",
       "Q61R                               3\n",
       "T58I                               3\n",
       "E17K                               3\n",
       "Q61L                               3\n",
       "Q61H                               3\n",
       "K117N                              2\n",
       "Y42C                               2\n",
       "Promoter Hypermethylation          2\n",
       "Q209L                              2\n",
       "ETV6-NTRK3 Fusion                  2\n",
       "Q61K                               2\n",
       "A146V                              2\n",
       "R170W                              2\n",
       "M1R                                2\n",
       "EWSR1-ETV1 Fusion                  2\n",
       "I31M                               2\n",
       "Y64A                               2\n",
       "TMPRSS2-ETV1 Fusion                2\n",
       "T286A                              2\n",
       "G12S                               2\n",
       "G35R                               2\n",
       "P34R                               2\n",
       "Q22K                               2\n",
       "S222D                              2\n",
       "...                              ...\n",
       "Y155C                              1\n",
       "D404G                              1\n",
       "Y35C                               1\n",
       "R156C                              1\n",
       "S1206Y                             1\n",
       "A1022E                             1\n",
       "E462G                              1\n",
       "I1171T                             1\n",
       "I157T                              1\n",
       "N116H                              1\n",
       "R897Q                              1\n",
       "D402Y                              1\n",
       "S723F                              1\n",
       "V379I                              1\n",
       "E709K                              1\n",
       "D1091N                             1\n",
       "N581S                              1\n",
       "A636P                              1\n",
       "C24Y                               1\n",
       "R625L                              1\n",
       "A1066V                             1\n",
       "L1301R                             1\n",
       "K459_S460delinsN                   1\n",
       "L858Q                              1\n",
       "P179L                              1\n",
       "L485F                              1\n",
       "Y65C                               1\n",
       "W557_V560del                       1\n",
       "S240R                              1\n",
       "STRN-PDGFRA Fusion                 1\n",
       "\n",
       "[2996 rows x 1 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(train_var['Variation'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# I want to find out which gene the variations with only one occurance below to\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7    953\n",
       "4    686\n",
       "1    568\n",
       "2    452\n",
       "6    275\n",
       "5    242\n",
       "3     89\n",
       "9     37\n",
       "8     19\n",
       "Name: Class, dtype: int64"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_var['Class'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAETCAYAAAAs4pGmAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAG5dJREFUeJzt3Xu0HGWZ7/HvzoWrAaIEUGREdHgGUe43ucYZBUEUxwF1\nKQ4XQdAgDDqDcnNUbioIguigEQw3HSV4A4zoQcCACILgAYFHuRyP64hjxABBCJBknz/e2qTZvMnu\nhOyunezvZ629VndVddXT1b371+/7VlX39ff3I0nSYGPaLkCSNDIZEJKkKgNCklRlQEiSqgwISVKV\nASFJqhrXdgFaNiJiQ+B+4M5m0hjgGeDszLyoWebTwH0D9xexnk8Av87M71fmPfv4iOgHJmXmX5ag\nxm2B92fm4RGxDfDxzNy328cvjYgYC3wH2AQ4JzPPHc7tLaaOQ4CVMvPLEXE4sFZmfqaH238LsH1m\nfqLb5SLibcAbM/PIYaxrQ+CuzHzRcG1DS8+AWLE8mZlbDNyJiFcA10TE3zLz8qE+HBr/CNxdm9Hl\n4xdnU+DlzbpuBYY1HBrrA3sAq2fm/B5sb1F2Bu4CyMzzWtj+tsCLl2S5zPwB8IPhLEojW58nyq0Y\nFvVNLCLeAxyZmTtExLRmmTMi4lPAPwNPAw8DBwLvAD4LzAI+AuxD+bB4FXAlsG7H4/uBr1I+UMYA\nJ2TmlRFxILBvZu7dbP9AShB8ELgRWJPyjf5C4NzMfG1ErAl8CdgC6AdmAMdl5ryImAt8BngT8DJK\ni+gLlee/C3A6sFrznE5otnczEJSW1b9k5v0dj5kEfL15fg8Df2qe3ycjYhPgbOAlwFhK6+OCiJgM\nnAI8ALwWWBmYkpnXRsRKzf7brXnM7cCRwD8B5wNPAqcCk4C1M/OIiNgY+AqwDrAAODkzvzXouX2y\nqfFVzT64GfgxcADwSuCYzPxms9zamXlEx+PWBi4Gvt/U9NWmhv8CNm5e3znAe4C1Bi33u4HXMiJe\n3jxmQ6APuDAzT2/ed9cAPwS2b9Z3/ODn0NSzN3Ay5f3yN+Bw4NFmn78oItZt9sW6wHrA74F3Zuaf\nI+KDzfJPA3OBwzLz7sVMXx84F/g7YDzw35l5akSMA75ICeynm9fxoMx8fHC9cgxiNPg18LrOCRGx\nAfBvwLaZuQ3lw2b7zPwScCvwH5n53Wbx1TJz08z8WGXdD2TmVsD+wIXNB25VZv4B+AQwMzMPGjT7\nHMoH9OuAbYDNgX9v5q0M/CUzd6IEzWciYpVBz+clwHTgqMzcjPLBeQnlw3EvmpZVZzh0bPc3mbkJ\nsB+wY7O+cc36Pp6ZW1M+8P89InZoHrc98PnM3JLywf/JZvrHgXnA1pm5OfBH4DPNvvwBcFazjzv9\nN3BZZm7a1HpqRKxR2YU7A3tSusreBLwmM3cFjgA+VVn+WZl5M3Ae8K3MPL5ZzyOZuUNmbgz8Ejii\nslynS4FrM/N1wE7A/hHx7mbeRsDVmbkd8DHgc4NraD78LwEObF6j0ynB3+ndwE2Z+fpmnU8A72u6\nCb8AvDkzt6WE186Lmt6s62Lggub12w54Y0S8E3g9MBnYrJn3ALDZ4vbfaGZArPj6Kf9onf4fJTh+\nFRFnAHdk5vcW8fgbFrPu8wAy8y5Kt9Trl7LGPSmtif7MfKpZ754d8wfGQ35FCYzVBz1+e8rYyM1N\nPb+htB4mD7HdvSgfKmTmQ5RQgPLN+lXABRFxB3A9sCqwZTP/95l5R0dNA103e1NaXbc3j3s78JpF\nbTwiXkwJw681NfwhM1+VmY9VFv9fmfloZj5JCZ4fNdPvp7uuo2dl5nRgWkR8OCLOpuynRY4BRMTq\nlFD4UvP4R4FpLHyNnqG0IOC5+6PTTpSWwh3NOr6TmZ2vMZl5NvDziPgI8GVKC+1FTdfgZc28cymt\njvMXNb2pdzfgpOZ1+AWlJbEFpSU5H7g5Ik4CLs/Mnw+910YnA2LFty0LB64ByMwFlH+gAynf3M9q\nPihqFtf07uzT76N8UPQ3twes1EWNg9+HYyjdAgOeBMjMgf7QvsrytXWOr0zvNG/Qugaez1jKN+wt\nBv6AHSjdUc/W0+h8vmMprZiBx2zH4sdZ5nWsA4AoVq0s+9Sg+89Ululq3zfdMudTvjh8A/gmz9+n\nncZU5nfu36eb91SthgHzeO7z7IuI53xzj4jPAp+mdHF+ldKy7QPIzP2BtwL3UVop31nM9LHN43Yc\n9PqdmpmPsLCFOh/4VkQcvZjnPqoZECuwpn/7RODzg6ZvThkwvSczTwPOovzTQPlHHuqDdcCBzfq2\nAv6e0jc+C3htRKzSdNW8tWP5Ra37amBK86GxMvAB4Cdd1gDlG2JExHZNPZsCuwLXDfG4q4D3N495\nCWVMph9IYG5E7N/M24Cyv7YeYn1XA0dExEoRMQaYCpzWzHvec29aCrdRusQGtjMwTrM0ZgFbN/tx\ndWD3jnmd298DmJaZ51Oe61spH6qLqnMOZR9PaepcE/hXluw1uhnYpHltoLS0Lhm0zB7AFzLzYuDP\nlK60sRGxdkT8AXi4GX86Adh8UdOb/foLyjgaEbEWZb/u04yDXAP8PDM/CVzEwve+BjEgViyrRsQd\nzd+vKN0Ax2bmVZ0LZeavgW8Dt0bErcDBwMC3qCuAMyLigC62t1FE3E7pInl3Zv6V8q3veuBeYCbP\nbb3cBPxDRHx30HqOpAzS3tn8JWUguCvNobb7AV+MiDsp34oPyszfDvHQo5t67gQupwyKPpGZT1M+\nwA6JiP/dPKcTM/PGIdZ3EvB/KIPTd1O+xX60mTcDODIijh30mPcA74yIX1P2/SGZ+aehnvMiXEoJ\nid9Runxu6ph3DfC2iPgicAZwWNP9cg2lW+jVleU6vRf4p2Zf3ULZX9O6LSwz/6dZx4XNdj9CGXPo\n9GnKe+82SkvgBuDVzet7MuWIvNsoYxeHLGp6s673ADs09d4MfDMzL6W8Dr8B7mre+zuycAxJg3gU\nk0atiPgQcHtm3tS0XGYC/5mZM1ouTRoRPA9Co9ndlFbHWEp//WWGg7SQLQhJUtWwtiAiYnvgs5k5\nOSJeTemz7KcM+E3JzAURcShwGGVw7OTmZKtVKQNY61BO4jkgM2cNZ62SpOcatkHqiDiGMng5cFLT\nmZSzbXehDN7tExHrUQYod6IcwXBa0xf8QeDOZtmLKEcnSJJ6aDhbEPdTLt1wcXN/a8rRLVCOJNid\nchzyjc3JUU9FxH2Usxp3ZuHZmDMoh2oOadasOfaXSdISmjRpQvU8mGELiMy8vLlOy4C+jhOd5lCO\n9V6DcvYji5k+MG1IEyeuxrhxY4deUJI0pF4exbSg4/YE4BHgseb24qYPTBvS7NmDryghSRrKpEkT\nqtN7eaLc7c2VMKFcw2Um5YSbXZqzbtekXIjsLspZj3sNWlaS1EO9DIiPAp+KiJsox5xPb84YPYcS\nAD+lXCZ4LuWywptGxA2Uyy4s9mqVkqRlb4U6D8JBaklacosapPZaTJKkKgNCklRlQEiSqgwISVKV\nASFJqvJy35KWuZNu6Orc1mXqxJ3X6vk2V3S2ICRJVQaEJKnKgJAkVRkQkqQqA0KSVGVASJKqDAhJ\nUpUBIUmqMiAkSVUGhCSpyoCQJFUZEJKkKgNCklRlQEiSqgwISVKVASFJqjIgJElVBoQkqcqAkCRV\nGRCSpCoDQpJUZUBIkqoMCElSlQEhSaoyICRJVQaEJKnKgJAkVRkQkqQqA0KSVDWulxuLiPHAhcCG\nwHzgUGAeMA3oB+4CpmTmgog4FDismX9yZl7Zy1olabTrdQtiL2BcZu4IfBo4BTgTOCEzdwH6gH0i\nYj3gSGAnYA/gtIhYuce1StKo1uuA+C0wLiLGAGsAzwBbA9c382cAbwS2A27MzKcy81HgPmCzHtcq\nSaNaT7uYgMcp3Uv3AmsDewO7ZmZ/M38OsCYlPB7teNzA9MWaOHE1xo0buyzrlbRUHun5FidNmtDz\nba7oeh0QRwNXZ+axEbEB8FNgpY75EyjvrMea24OnL9bs2U8sw1IlLU9mzZrTdgnLrUWFa6+7mGaz\nsGXwV2A8cHtETG6m7QnMBG4BdomIVSJiTWATygC2JKlHet2COAu4ICJmUloOxwG3AlMjYiXgHmB6\nZs6PiHMoYTEGOD4z5/a4Vkka1fr6+/uHXmo5MWvWnBXnyUjLsZNu6P0YxIk7r9Xzba4oJk2a0Feb\n7olykqQqA0KSVGVASJKqDAhJUpUBIUmqMiAkSVUGhCSpyoCQJFUZEJKkKgNCklRlQEiSqgwISVKV\nASFJqjIgJElVBoQkqcqAkCRVGRCSpCoDQpJUZUBIkqoMCElS1bi2C5Bqjrlhv1a2+7mdL2tlu9JI\nZAtCklRlQEiSqgwISVKVASFJqjIgJElVBoQkqcqAkCRVGRCSpCoDQpJUZUBIkqoMCElSlQEhSaoy\nICRJVQaEJKnKgJAkVfX89yAi4ljgbcBKwJeB64FpQD9wFzAlMxdExKHAYcA84OTMvLLXtUrSaNZV\nCyIifhgR+0XE+BeysYiYDOwI7ATsBmwAnAmckJm7AH3APhGxHnBks9wewGkRsfIL2bYkacl024L4\nDHAAcHpEXAVMy8xfLsX29gDuBL4LrAH8B3AopRUBMAPYHZgP3JiZTwFPRcR9wGbAYrc5ceJqjBs3\ndinKkopJkya0XcIK4pGeb9HXbtnrKiAy82fAzyJiVWBf4PKIeAz4GvBfzQd5N9YGXgHsDbwS+AEw\nJjP7m/lzgDUp4fFox+MGpi/W7NlPdFmGVDdr1py2S9BS8rVbeosK164HqZvuoXOBU4EfAUcB61E+\n5Lv1MHB1Zj6dmQnM5bkf/BMoXz0ea24Pni5J6pGuWhAR8XvgAeDrwBGZ+WQz/TqG6PYZ5AbgqIg4\nE3gpsDpwTURMzszrgD2Ba4FbgFMiYhVgZWATygC2JKlHuh2D+EdgTmb+OSJWjYhXZ+Z9mTkf2Krb\njWXmlRGxKyUAxgBTgAeBqRGxEnAPMD0z50fEOcDMZrnjM3PuEjwvSdIL1G1AvAU4kBIG6wBXRMRZ\nmfnVJd1gZh5TmbxbZbmpwNQlXb8kadnodgziA8AuAJn5e2Br4MPDVZQkqX3dtiDGA51HKj1NObFt\n5Jr+/Xa2u+8+7WxXkpaxbgPie8BPI+Lbzf13sGRHL0mSljNddTFl5seAc4AANgLOycwThrMwSVK7\nluRiffcA36a0Jv7aHI0kSVpBdXsexJeAtwL3d0zupxz+KklaAXU7BrE7EAMnyEmSVnzddjE9QLnS\nqiRplOi2BfFX4O6I+Dnl+kkAZObBw1KVJKl13QbEj5o/SdIo0e3lvi+MiA2BTYGrgQ0y88HhLEyS\n1K5uf1HuXcAVwNnAi4GbImL/4SxMktSubgepP0b5qdA5mflnYEvg2GGrSpLUum4DYn5mPvtzTZn5\nELBgeEqSJI0E3Q5S/yYijgDGR8QWwIeAO4avLElS27ptQUwB1geeBC6g/CToh4arKElS+7o9iulv\nlDEHxx0kaZTo9lpMC3j+7z88lJkvX/YlSZJGgm5bEM92RUXEeODtwOuHqyhJUvuW5HLfAGTmM5l5\nGV7JVZJWaN12Mf1rx90+yhnVTw9LRZKkEaHbw1zf0HG7H/gL8K5lX44kaaTodgzioOEuRJI0snTb\nxfQgzz+KCUp3U39mbrRMq5Ikta7bLqZvAE8BU4FngPcC2wLHD1NdkqSWdRsQe2TmNh33z46I2zLz\n98NRlCSpfd0e5toXEW8cuBMRe1MutyFJWkF124L4AHBRRKxHGYu4Fzhg2KpagT00/b093+ZL9720\n59uUtPzr9iim24BNI2JtYG5mPj68ZUmS2tbtL8q9IiJ+AtwEvCgiftr8BKkkaQXV7RjEV4DTgceB\n/wG+CVw0XEVJktrXbUCsnZk/BsjM/sycCqwxfGVJktrWbUA8GREvpzlZLiJ2ppwXIUlaQXV7FNPR\nwJXAqyLiDuDFwH7DVpUkqXXdBsS6lDOnNwbGAvdmpldzlaQVWLcB8bnMvAr4zbLYaESsA9wGvAmY\nB0yjdF/dBUzJzAURcShwWDP/5My8cllsW5LUnW4D4v6IuAC4GXhyYGJmLvGRTM0v0n2lYz1nAidk\n5nURcR6wT0TcBBwJbAOsAtwQET/JTMc9JKlHFjtIHRHrNzcfply5dQfKb0O8AZi8lNs8AzgP+GNz\nf2vg+ub2DOCNwHbAjZn5VGY+CtwHbLaU25MkLYWhWhBXAFtl5kER8dHM/PwL2VhEHAjMysyrI+LY\nZnJfZg5cSnwOsCblENpHOx46MH2xJk5cjXHjxgIw64UU+gJMmjRhsfMf6lEdnYaqSQu5r5aVR3q+\nRV+7ZW+ogOjruP1e4AUFBHAw0N9c+G8Lysl263TMn0B5Zz3W3B48fbFmz37iBZb3ws2aNaftEp5n\nJNY0Urmvll++dktvUeE6VEB0/khQ3yKX6lJm7jpwOyKuAw4HTo+IyZl5HbAncC1wC3BKRKwCrAxs\nQhnAliT1SLeD1FD/Rbll4aPA1IhYCbgHmJ6Z8yPiHGAmZZzk+MycO0zblyRVDBUQm0bEA83t9Ttu\nv+CfGs3MyR13d6vMn0r5BTtJUguGCoiNe1KFJGnEWWxA+JOikjR6dXuxPknSKGNASJKqDAhJUpUB\nIUmqMiAkSVUGhCSpyoCQJFUZEJKkKgNCklRlQEiSqgwISVKVASFJqjIgJElVBoQkqcqAkCRVGRCS\npCoDQpJUZUBIkqoMCElSlQEhSaoa13YBkpbe+6+/vpXtnr/bbq1sV71lC0KSVGVASJKqDAhJUpUB\nIUmqMiAkSVUGhCSpysNcpS4dOPPiVrY7bZf3tbJdyRaEJKnKgJAkVRkQkqQqA0KSVGVASJKqDAhJ\nUlVPD3ONiPHABcCGwMrAycDdwDSgH7gLmJKZCyLiUOAwYB5wcmZe2ctaJWm063ULYn/g4czcBXgz\ncC5wJnBCM60P2Cci1gOOBHYC9gBOi4iVe1yrJI1qvT5R7jJgenO7j9I62BoYuKj9DGB3YD5wY2Y+\nBTwVEfcBmwG/XNzKJ05cjXHjxgIwa5mX3p1JkyYsdv5DPaqj01A1aaGRuK+Wz5oe6UkdnUbiflre\n9TQgMvNxgIiYQAmKE4AzMrO/WWQOsCawBvBox0MHpi/W7NlPLNN6l8asWXPaLuF5RmJNI9VI3FfW\n1J2RWNPyYlHh2vNB6ojYALgWuDgzvwEs6Jg9gfLV47Hm9uDpkqQe6WlARMS6wI+Bj2XmBc3k2yNi\ncnN7T2AmcAuwS0SsEhFrAptQBrAlST3S6zGI44CJwIkRcWIz7SjgnIhYCbgHmJ6Z8yPiHEpYjAGO\nz8y5Pa5Vkka1Xo9BHEUJhMGe9wvomTkVmDrsRUmSqjxRTpJUZUBIkqoMCElSlQEhSaoyICRJVQaE\nJKnKgJAkVRkQkqQqA0KSVGVASJKqDAhJUpUBIUmqMiAkSVUGhCSpyoCQJFUZEJKkKgNCklRlQEiS\nqnr9m9QagX501b6tbPfNb5neynYldccWhCSpyoCQJFUZEJKkKgNCklRlQEiSqgwISVKVASFJqjIg\nJElVBoQkqcqAkCRVGRCSpCqvxSRJLVlwyYOtbHfM/q/sbrlhrkOStJwyICRJVQaEJKnKMQhJo8Jv\nZ8xtZbsb77lKK9tdFkZsQETEGODLwObAU8AhmXlfu1VJ0ugxkruY3g6skpmvBz4OfL7leiRpVBnJ\nAbEz8COAzPwFsE275UjS6NLX39/fdg1VEfE14PLMnNHc/7/ARpk5r93KJGl0GMktiMeACR33xxgO\nktQ7IzkgbgT2AoiIHYA72y1HkkaXEXsUE/Bd4E0R8XOgDzio5XokaVQZsWMQkqR2jeQuJklSiwwI\nSVKVASFJqhrJg9StiYjtgc9m5uQRUMt44AJgQ2Bl4OTM/EHLNY0FpgIB9AOHZ+ZdbdY0ICLWAW4D\n3pSZ97ZdD0BE/Ipy2DbAg5nZ+gEXEXEs8DZgJeDLmXl+y/UcCBzY3F0F2AJYLzMfabGm8cCFlP+9\n+cChbb+nImJl4OvARpT31JTM/N1wbc8WxCARcQzwNcqbdCTYH3g4M3cB3gyc23I9AG8FyMydgBOA\nU9otp2j+ob8CPNl2LQMiYhWgLzMnN38jIRwmAzsCOwG7ARu0WhCQmdMG9hEl4I9sMxwaewHjMnNH\n4NOMjPf5ocDjmbkD8GGG+fPAgHi++4F3tF1Eh8uAE5vbfUDrJwtm5veADzR3XwG0/Y884AzgPOCP\nbRfSYXNgtYj4cUT8tDmnp217UM4r+i5wBXBlu+UsFBHbAJtm5lfbrgX4LTCuuXDoGsAzLdcD8Bpg\nBkBmJrDJcG7MgBgkMy9nZLwRAMjMxzNzTkRMAKZTvrG3LjPnRcSFwBeBS9uup+mimJWZV7ddyyBP\nUIJrD+Bw4NKIaLtrd23Ktc32Y2FNfe2W9KzjgE+1XUTjcUr30r2ULtVzWq2muAPYOyL6mi8b6zdd\nvsPCgFgORMQGwLXAxZn5jbbrGZCZBwAbA1MjYvWWyzmYcmLldZT+64siYr12SwLKt9BLMrM/M38L\nPAy8tOWaHgauzsynm2+hc4FJLddERKwFRGZe23YtjaMp+2ljSkvwwqbLsE0XUMYeZgL/DNyWmfOH\na2Ntf5PRECJiXeDHwBGZeU3b9QBExPuAl2fmaZRvyAuav9Zk5q4Dt5uQODwz/9ReRc86GHgd8KGI\neBmlq+KhdkviBuCoiDiTElarU0KjbbsCI+I93pjNwt6EvwLjgWH7tt6lbYFrMvPopjvuFcO5MQNi\n5DsOmAicGBEDYxF7ZmabA7HfAb4eET+j/NP8W8v1jGTnA9Mi4gbKEV8Ht33Rycy8MiJ2BW6h9CJM\nGc5voUsggAfaLqLDWcAFETGTcrTXcZn5t5Zr+h1wUkQcTxn7e/9wbsxLbUiSqhyDkCRVGRCSpCoD\nQpJUZUBIkqoMCElSlYe5SkshItYATqNcy2ge5Zj5j1LOc/jkSLjQo/RC2YKQllBzbZ4fUk6e2iIz\nt6BczG0G8JI2a5OWJVsQ0pJ7A/Ay4D8zcwFAZl4bEQcBLxpYKCJ2o1wBdDXKyY7HZOZlEfEe4BjK\nJaQfpFyxd23KNa1Wp5yVfmRm/qJ3T0l6PlsQ0pLbEvjlQDgMyMwfAn/umPRh4JDM3Ipyxusnmukn\nA7tn5taUC8H9QzP/yszchhIeOw/vU5CGZgtCWnILKJdeH8r+lCtv7gfswMLWxRXAjRHxPeDyzLyj\nudjhdyJiS+AqRsbvfmiUswUhLblbga0GXyI7Ik7lucExE9iO8gM4pwzMy8yjgH+hjGFcEhH7Z+aN\nlGv9Xw28ixIiUqsMCGnJzaR0Jf3nwLX4I2IP4CBgneb+iymXQv9E0/W0OzA2IsZFxO+AvzRXw70I\n2DIiPge8LzMvBI4Atur1k5IG82J90lKIiLUpV/vchnJJ6L9QDnNdk+Yw14j4PPB2yvX7b6K0DP6O\n8lvQJ1Iulf4IcADly9o3gAmUwevPZua3e/mcpMEMCElSlV1MkqQqA0KSVGVASJKqDAhJUpUBIUmq\nMiAkSVUGhCSp6v8DerlWBzamq3UAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x19751c978>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.figure()\n",
    "sns.countplot(x=\"Class\", data=train_var)\n",
    "plt.ylabel('Frequency')\n",
    "plt.xlabel('Class')\n",
    "plt.title(\"Distribution of genetic mutation classes\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Looking at the genes with 50 or more occurances in the training data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5    90\n",
      "1    60\n",
      "6    53\n",
      "4    46\n",
      "3    15\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# BRCA1\n",
    "BRCA1 = train_var.loc[train_var['Gene']=='BRCA1']\n",
    "print(BRCA1['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1    83\n",
      "4    67\n",
      "6     5\n",
      "2     3\n",
      "5     2\n",
      "3     2\n",
      "9     1\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# TP53\n",
    "TP53 = train_var.loc[train_var['Gene']=='TP53']\n",
    "print(TP53['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7    84\n",
      "2    45\n",
      "4     5\n",
      "5     3\n",
      "8     1\n",
      "6     1\n",
      "3     1\n",
      "1     1\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# EGFR\n",
    "EGFR = train_var.loc[train_var['Gene']=='EGFR']\n",
    "print(EGFR['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4    115\n",
      "3      5\n",
      "1      3\n",
      "5      2\n",
      "7      1\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# PTEN\n",
    "PTEN = train_var.loc[train_var['Gene']=='PTEN']\n",
    "print(PTEN['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6    84\n",
      "5    20\n",
      "1    18\n",
      "4     3\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# BRCA2\n",
    "BRCA2 = train_var.loc[train_var['Gene']=='BRCA2']\n",
    "print(BRCA2['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7    52\n",
      "2    44\n",
      "6     1\n",
      "4     1\n",
      "3     1\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# KIT\n",
    "KIT = train_var.loc[train_var['Gene']=='KIT']\n",
    "print(KIT['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7    48\n",
      "2    31\n",
      "5     8\n",
      "6     3\n",
      "4     2\n",
      "3     1\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# BRAF\n",
    "BRAF = train_var.loc[train_var['Gene']=='BRAF']\n",
    "print(BRAF['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7    30\n",
      "2    14\n",
      "6    12\n",
      "5     6\n",
      "4     5\n",
      "8     1\n",
      "1     1\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# ERBB2\n",
    "ERBB2 = train_var.loc[train_var['Gene']=='ERBB2']\n",
    "print(ERBB2['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7    47\n",
      "3    10\n",
      "2     7\n",
      "5     5\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# ALK\n",
    "ALK = train_var.loc[train_var['Gene']=='ALK']\n",
    "print(ALK['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7    42\n",
      "3     9\n",
      "5     4\n",
      "2     3\n",
      "4     1\n",
      "1     1\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# PDGFRA\n",
    "PDGFRA = train_var.loc[train_var['Gene']=='PDGFRA']\n",
    "print(PDGFRA['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7    31\n",
      "2    14\n",
      "5     8\n",
      "6     3\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# PIK3CA\n",
    "PIK3CA = train_var.loc[train_var['Gene']=='PIK3CA']\n",
    "print(PIK3CA['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4    47\n",
      "1     4\n",
      "6     1\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# CDKN2A\n",
    "CDKN2A = train_var.loc[train_var['Gene']=='CDKN2A']\n",
    "print(CDKN2A['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7    27\n",
      "2     7\n",
      "1     7\n",
      "5     5\n",
      "6     4\n",
      "Name: Class, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# FGFR2\n",
    "FGFR2 = train_var.loc[train_var['Gene']=='FGFR2']\n",
    "print(FGFR2['Class'].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Training Text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3321, 2)\n",
      "ID      0\n",
      "Text    0\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung canc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1   Abstract Background  Non-small cell lung canc...\n",
       "2   2   Abstract Background  Non-small cell lung canc...\n",
       "3   3  Recent evidence has demonstrated that acquired...\n",
       "4   4  Oncogenic mutations in the monomeric Casitas B..."
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(train_txt.shape)\n",
    "print(train_txt.isnull().sum())\n",
    "train_txt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# len(train_txt['Test'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "### Test Variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5668, 3)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ACSL4</td>\n",
       "      <td>R570S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>NAGLU</td>\n",
       "      <td>P521L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>PAH</td>\n",
       "      <td>L333F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ING1</td>\n",
       "      <td>A148D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TMEM216</td>\n",
       "      <td>G77A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID     Gene Variation\n",
       "0   0    ACSL4     R570S\n",
       "1   1    NAGLU     P521L\n",
       "2   2      PAH     L333F\n",
       "3   3     ING1     A148D\n",
       "4   4  TMEM216      G77A"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(test_var.shape)\n",
    "test_var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5668, 2)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>2. This mutation resulted in a myeloproliferat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract The Large Tumor Suppressor 1 (LATS1)...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Vascular endothelial growth factor receptor (V...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Inflammatory myofibroblastic tumor (IMT) is a ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Abstract Retinoblastoma is a pediatric retina...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  2. This mutation resulted in a myeloproliferat...\n",
       "1   1   Abstract The Large Tumor Suppressor 1 (LATS1)...\n",
       "2   2  Vascular endothelial growth factor receptor (V...\n",
       "3   3  Inflammatory myofibroblastic tumor (IMT) is a ...\n",
       "4   4   Abstract Retinoblastoma is a pediatric retina..."
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(test_text.shape)\n",
    "test_txt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID       int64\n",
      "Text    object\n",
      "dtype: object\n",
      "ID      0\n",
      "Text    0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(train_txt.dtypes)\n",
    "print(train_txt.isnull().sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Feature Selection\n",
    "Not sure what I will need to do for feature selection because all of the columns are important here. It maybe the case here that I will need to add aditional features as outlined above in the approved outside resources section. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Modeling\n",
    "I will make use of Natural Language Processing methods such as Bag of Words and Tf-idf. \n",
    "\n",
    "I will be able to do more in this section after lecture Monday and Tuesday."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
